### PR TITLE
SASRec Experiments

### PR DESCRIPTION
**Pull request by haru-256**
## Summary
- conducted experiment of SASRec
- There are TODO
  - Eval ranking metrics (e.g. NDCG, MRR) for validation.

___

### **PR Type**
enhancement, bug fix, tests

___

### **Description**
- Enhanced the SASRec model with support for float16 training.
- Improved validation data handling and added NaN detection in attention layers.
- Updated README with detailed model architecture and results.
- Added utility functions and optimized training configurations.

___

### **Changes walkthrough** 📝
<table><thead><tr><th></th><th align="left">Relevant files</th></tr></thead><tbody><tr><td><strong>Enhancement</strong></td><td><details><summary>6 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>dataset.py</strong><dd><code>Update validation data loader</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

sasrec/data/dataset.py

<li>Changed the data loader to use <code>val_dataset</code> instead of <code>test_dataset</code>.<br>

</details>

  </td>
  <td><a href="https://github.com/haru-256/ml-sandbox/pull/4/files#diff-2be12967692ffa8605fb196e2eeac2525882bb613bc7b78c8b354f4074a36184">+1/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>sasrec.py</strong><dd><code>Add float16 support for model training</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

sasrec/models/sasrec.py

<li>Added <code>float16</code> parameter for model optimization.<br> <li> Updated attention mask creation to support <code>float16</code>.<br>

</details>

  </td>
  <td><a href="https://github.com/haru-256/ml-sandbox/pull/4/files#diff-744591cb834f7c304aef7065215ac5234d64418f85583b5e382011178aed49ac">+16/-11</a>&nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>train.py</strong><dd><code>Optimize training configuration</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

sasrec/train.py

<li>Increased batch size for training.<br> <li> Configured model to use GPU if available.<br>

</details>

  </td>
  <td><a href="https://github.com/haru-256/ml-sandbox/pull/4/files#diff-8cf036aad727d6d2a7352c5dabc033e21c89aca0a12383456332f5411c96ca80">+15/-4</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>utils.py</strong><dd><code>Add utility function for CPU core count</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

sasrec/utils/utils.py

- Added `cpu_count` function to get the number of CPU cores.

</details>

  </td>
  <td><a href="https://github.com/haru-256/ml-sandbox/pull/4/files#diff-c56c3c1e23840b5175e31e86972f2386f0336194ff44d9fab1ae29a96e1e4311">+31/-3</a>&nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>Makefile</strong><dd><code>Update Makefile for training command</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

sasrec/Makefile

- Added `train` target to Makefile for easier execution.

</details>

  </td>
  <td><a href="https://github.com/haru-256/ml-sandbox/pull/4/files#diff-76a14341f4ffe31b5e7acddcc5edb2815cdd1eb633f55ba5534372f9b1054f21">+4/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>pyproject.toml</strong><dd><code>Update project dependencies</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

sasrec/pyproject.toml

- Updated dependencies for cross-platform support.

</details>

  </td>
  <td><a href="https://github.com/haru-256/ml-sandbox/pull/4/files#diff-06499470d8ce6335725a36f7507004d6425c1aa67a835f3e19d34437ff1ab8f1">+4/-1</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    
</table></details></td></tr><tr><td><strong>Bug fix</strong></td><td><details><summary>3 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>transformer_embedding.py</strong><dd><code>Fix embedding layer issues</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

sasrec/models/modules/transformer_embedding.py

<li>Fixed the embedding lookup to use <code>id_embeddings</code>.<br> <li> Updated position IDs to move to the correct device.<br>

</details>

  </td>
  <td><a href="https://github.com/haru-256/ml-sandbox/pull/4/files#diff-14092cb61aabea4611739da21e8512457931025291b46d54ee643470d410f0f2">+2/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>transformer_encoder_block.py</strong><dd><code>Enhance NaN detection in attention block</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

sasrec/models/modules/transformer_encoder_block.py

- Improved NaN detection in MultiheadAttention output.

</details>

  </td>
  <td><a href="https://github.com/haru-256/ml-sandbox/pull/4/files#diff-83ed5e1608b87362207d144fa715a0342d33822b8454001d8e7b827758e1c7aa">+6/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    

<tr>
  <td>
    <details>
      <summary><strong>test_utils.py</strong><dd><code>Fix padding mask creation in tests</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

sasrec/tests/test_utils/test_utils.py

- Updated padding mask creation to use system minimum value.

</details>

  </td>
  <td><a href="https://github.com/haru-256/ml-sandbox/pull/4/files#diff-41aa4812782ef2a618ae099e5db3e4e58c34118125f6288f2dbad609f2eb3db4">+3/-2</a>&nbsp; &nbsp; &nbsp; </td>

</tr>                    
</table></details></td></tr><tr><td><strong>Documentation</strong></td><td><details><summary>1 files</summary><table>
<tr>
  <td>
    <details>
      <summary><strong>README.md</strong><dd><code>Improve documentation with detailed model info</code>&nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; </dd></summary>
<hr>

sasrec/README.md

<li>Expanded dataset description and model architecture.<br> <li> Added results and usage instructions.<br>

</details>

  </td>
  <td><a href="https://github.com/haru-256/ml-sandbox/pull/4/files#diff-4c20b297e0aa35b6e103930e1e3e753ea21653efa0af38244b569befa5e182d1">+108/-1</a>&nbsp; </td>

</tr>                    
</table></details></td></tr></tr></tbody></table>

___

> 💡 **PR-Agent usage**:
>Comment `/help` on the PR to get a list of all available PR-Agent tools and their descriptions

### CODE REVIEW

File: sasrec/utils/utils.py [hunk 1]
```diff
@@ -1,28 +1,56 @@
...
+    attn_mask = attn_mask.to(device)
 
-    padding_mask = x == pad_idx  # shape (batch_size, seq_len)
-    padding_mask = torch.masked_fill(torch.zeros(batch_size, seq_len), padding_mask, float("-inf"))
+    padding_mask_bool = x == pad_idx  # shape (batch_size, seq_len)
+    # for example, if all elements in the rows of lower triangular matrix is equal to pad_idx, nn.MultiheadAttention will return NaN
+    # to prevent this, we set padding_mask to minimum value of float
+    # https://github.com/pytorch/pytorch/issues/24816
+    padding_mask = torch.masked_fill(
+        torch.zeros(batch_size, seq_len).to(device),
```
Comment (haru-256) [inline sasrec/utils/utils.py:36]:
```suggestion
        torch.zeros_like(x, dtype=torch.float32),
```
